pkg/resources/cloudcontroller/openstack.go

```diff
@@ -171,13 +171,19 @@ func getOSFlags(data *resources.TemplateData) []string {
 }
 
 func getOSVersion(version semver.Semver) (string, error) {
+	if version.Major() < 17 {
+		return "", fmt.Errorf("Kubernetes version %s is not supported", version.String())
+	}
+
 	switch version.Minor() {
 	case 17:
 		return "1.17.0", nil
 	case 18:
 		return "1.18.0", nil
+	case 19:
+		return "1.19.3", nil
 	default:
-		return "", fmt.Errorf("Kubernetes version %s is not supported", version.String())
+		return "1.19.3", nil
```

Comment:
Wouldn't be better to return an error in case we are trying to install a version greater than 1.19? We also need to change a bit the logic in this case. If you prefer we can do this in a follow up PR

Reply:
No, it was like that before this change, but it's not a good idea, IMO. It's better to fall-back to the latest minor version available than to fall-back to the in-tree cloud provider.

> We would also need to change a bit the logic to prevent the cluster from being created in such a case to avoid forgetting about this when we will introduce support for 1.20.

This is the biggest reason for this change. As stated above, it's much better to fall-back to the previous minor version, which is going to work perfectly fine even with a new Kubernetes version. It might happen that it'll take a bunch of time for the OpenStack CCM team to release a minor version to match the latest Kubernetes version and in such a scenario:

- we don't want to block adding new Kubernetes version support
- we want to ensure that users can still deploy OpenStack clusters (i.e. returning an error is not an option)
- we want to ensure that we don't fall-back to the in-tree CCM

Reply:
Sorry, I was not clear enough with my comment. I did not mean that we should fallback to in-tree cloud provider. I meant that we should block creating OpenStack cluster with minor version > `19` with 1.19 CCM. I think it's better to fail fast than succeeding in degraded mode (with old CCM in this case), because we risk missing this again (even if consequencese will be less important). 

I agree with you that returning an error with current logic will result in fall-back to in-tree, and this is definitely not what we want. That's why I also suggested we could handle this with a follow-up PR.

Reply:
I think that blocking cluster creation is not a good idea. For example, what if it takes weeks before the OpenStack CCM is available? Blocking cluster creation for that long would block both us and users from testing OpenStack with the latest Kubernetes version.

> I think it's better to fail fast than succeeding in degraded mode (with old CCM in this case)

It's not running in a degraded mode. It would have all the features that were available in the latest minor version available (for example 1.19). We have checked with the OpenStack CCM team and SIG Cloud Provider about this and it's totally okay to do that. For example, in KubeOne, we unconditionally deploy OpenStack CCM 1.17.0, and it works flawlessly well for any Kubernetes version.

Of course, if you want the latest features, you have to deploy the latest CCM, but until it's not available, it's much better to use the previous version rather than block users from deploying clusters.

> because we risk missing this again (even if consequencese will be less important).

This can be solved by creating an issue to remind us to add the latest version.

Reply:
Yep, but we have to remind to create the reminder each time :-). What you did is way better than what we have, but as we will be working explicitly on support for k8s `1.20`: having the OpenStack test failing is IMO a good reminder. It will force us to tackle the point. If no CCM will be available at this stage for 1.20 and the 1.19 will be proven to work fine we will use it.

Reply:
That said, I think this is good enough to fix. I'll think about it and maybe create a follow-up PR to propose what I have in mind.

Reply:
Sounds like a good plan. :+1: